builtins/web/blob.cpp

```diff
@@ -0,0 +1,703 @@
+    return JS::ToString(cx, value);
+  } else {
+    return JS_GetEmptyString(cx);
+  }
+
+  auto str = JS::StringToLinearString(cx, value_str);
+  if (!str) {
+    return nullptr;
+  }
+
+  std::string normalized;
+  auto strlen = JS::GetLinearStringLength(str);
+
+  if (!strlen) {
+    return JS_GetEmptyString(cx);
+  }
+
+  if (JS::LinearStringHasLatin1Chars(str)) {
+    JS::AutoCheckCannotGC nogc(cx);
+    auto chars = JS::GetLatin1LinearStringChars(nogc, str);
+    if (!validate_type(chars, strlen)) {
+      return JS_GetEmptyString(cx);
+    }
+
+    normalized = std::string(reinterpret_cast<const char *>(chars), strlen);
+  } else {
+    JS::AutoCheckCannotGC nogc(cx);
+    auto chars = (JS::GetTwoByteLinearStringChars(nogc, str));
+    if (!validate_type(chars, strlen)) {
+      return JS_GetEmptyString(cx);
+    }
+
+    normalized.reserve(strlen);
+    for (size_t i = 0; i < strlen; ++i) {
+      normalized += static_cast<unsigned char>(chars[i]);
+    }
+  }
+
+  std::transform(normalized.begin(), normalized.end(), normalized.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+
+  return JS_NewStringCopyN(cx, normalized.c_str(), normalized.length());
+}
+
+// https://w3c.github.io/FileAPI/#convert-line-endings-to-native
+std::string convert_line_endings_to_native(std::string_view s) {
+  std::string native_line_ending = "\n";
+#ifdef _WIN32
+  native_line_ending = "\r\n";
+#endif
+
+  std::string result;
+  std::regex re(R"(\r\n|\r|\n)");
+  std::regex_replace(std::back_inserter(result), s.begin(), s.end(), re, native_line_ending);
+
+  return result;
+}
+
+} // anonymous namespace
+
+namespace builtins {
+namespace web {
+namespace blob {
+
+const JSFunctionSpec Blob::static_methods[] = {
+    JS_FS_END,
+};
+
+const JSPropertySpec Blob::static_properties[] = {
+    JS_PS_END,
+};
+
+const JSFunctionSpec Blob::methods[] = {
+    JS_FN("arrayBuffer", Blob::arrayBuffer, 0, JSPROP_ENUMERATE),
+    JS_FN("bytes", Blob::bytes, 0, JSPROP_ENUMERATE),
+    JS_FN("slice", Blob::slice, 0, JSPROP_ENUMERATE),
+    JS_FN("stream", Blob::stream, 0, JSPROP_ENUMERATE),
+    JS_FN("text", Blob::text, 0, JSPROP_ENUMERATE),
+    JS_FS_END,
+};
+
+const JSPropertySpec Blob::properties[] = {
+    JS_PSG("size", size_get, JSPROP_ENUMERATE),
+    JS_PSG("type", type_get, JSPROP_ENUMERATE),
+    JS_STRING_SYM_PS(toStringTag, "Blob", JSPROP_READONLY),
+    JS_PS_END,
+};
+
+class StreamTask final : public api::AsyncTask {
+    Heap<JSObject *> source_;
+
+    static constexpr size_t CHUNK_SIZE = 8192;
+
+public:
+  explicit StreamTask(const HandleObject source, api::Engine *engine)  : source_(source) {
+    engine->incr_event_loop_interest();
+    handle_ = IMMEDIATE_TASK_HANDLE;
+  }
+
+  [[nodiscard]] bool run(api::Engine *engine) override {
+    JSContext *cx = engine->cx();
+    RootedObject owner(cx, streams::NativeStreamSource::owner(source_));
+    RootedObject controller(cx, streams::NativeStreamSource::controller(source_));
```

Comment:
Instead of getting the controller here and then using `JS::Call` for the `close` and `enqueue` operations below, you can use `JS::ReadableStreamClose` and `JS::ReadableStreamEnqueue` on `source_` directly, I think.

Reply:
I wasn't able to make this working. If I try to use Close/Enque directly this fails on some downcasting assertions. I think in order for this to work I need to have a reference to a `DefaultReadableStream` object, but I wouldn't know how to get this from `source_`.

Reply:
Oh I see; you're not storing the stream itself anywhere, so you don't have access to it here. And `NativeStreamSource::stream` isn't any help, because it assumes that the owner will be a `BodySource` :(

It might be easiest to fix that by storing the stream itself on the source explicitly. If you're up for it, you could do that, and instead remove the `Controller` slot, because that can be retrieved from the stream, but not the other way around.

The reason all of this is important is that otherwise content can change what's happening here by setting different functions for `ReadableStreamDefaultController.prototype.{close, enqueue}`, which we absolutely don't want.

Reply:
Right, the initialization order now is as follows:

```cpp
  JS::RootedObject native_source(cx, streams::NativeStreamSource::create(
    cx, self, JS::UndefinedHandleValue, stream_pull, stream_cancel));

  JS::RootedObject source_stream(cx, JS::NewReadableDefaultStreamObject(cx, native_source, nullptr, 0.0));
```

So what you're proposing is to move the initialization of the default stream to `NativeStreamSource::create` and store it in a slot there?

Reply:
I think that'd make sense, yes

Reply:
I feared that might be the case :( What I'd do from here is to either reduce the test case / add `console.log` statements to it to get more information, or to add debugging output using the `DBG` macro to the C++ code.

But yes, that all should be for next week. For now, I hope you have a great weekend!

Reply:
OK, I think I've figured this out. The issue was that I initialized `native-stream` slots after passing the instance handle to `NewReadableDefaultStream` constructor.